### PR TITLE
add distributed mode in fc layer

### PR DESCRIPTION
**Pull request by Yancey1989**
model parallelism for FC layer.

1. the principle of the distributed mode of FC layer:
sync_fc

1. the scalability of the distributed FC layer.

GPUs | fc_w.dims | GPU memory(peak) | ratio
-- | -- | -- | --
1 | [2^5, 2^26] | 21103MiB | 1
2 | [2^5, 2^26] | 12945MiB | 61.34%
4 | [2^5, 2^26] | 9091MiB | 43.07%
8 | [2^5, 2^26] | 6919MiB | 32.78%
8 | [2^5, 2^26 * 3] | 18705MiB | -

### CODE REVIEW

File: paddle/fluid/API.spec [hunk 1]
```diff
@@ -49,7 +49,7 @@ paddle.fluid.initializer.BilinearInitializer.__init__ ArgSpec(args=['self'], var
 paddle.fluid.initializer.MSRAInitializer.__init__ ArgSpec(args=['self', 'uniform', 'fan_in', 'seed'], varargs=None, keywords=None, defaults=(True, None, 0))
 paddle.fluid.initializer.force_init_on_cpu ArgSpec(args=[], varargs=None, keywords=None, defaults=None)
 paddle.fluid.initializer.init_on_cpu ArgSpec(args=[], varargs='args', keywords='kwds', defaults=None)
-paddle.fluid.layers.fc ArgSpec(args=['input', 'size', 'num_flatten_dims', 'param_attr', 'bias_attr', 'act', 'is_test', 'name'], varargs=None, keywords=None, defaults=(1, None, None, None, False, None))
+paddle.fluid.layers.fc ArgSpec(args=['input', 'size', 'num_flatten_dims', 'param_attr', 'bias_attr', 'act', 'is_test', 'name', 'distributed', 'trainers'], varargs=None, keywords=None, defaults=(1, None, None, None, False, None, False, 1))
```
Comment (typhoonzero) [inline paddle/fluid/API.spec]: Can use a new layer to represent distributed fc with activation.

File: paddle/fluid/operators/sync_fc_allgather_op.cc [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ public:
+  void Make() override {
+    AddInput("X", "(LoDTensor) Input tensor of sync fc allgather operator.");
+    AddOutput("Out",
+              "(LoDTensor) Output tensor of sync fc allgather operator.");
+    AddAttr<int>("trainers",
+                 "Total trainer count of the distributed training job.")
+        .SetDefault(1);
+    AddComment(R"DOC(
+All gather operator of sync fc operator
```
Comment (typhoonzero) [inline paddle/fluid/operators/sync_fc_allgather_op.cc:105]: doc is too simple

File: paddle/fluid/platform/nccl_helper.h [hunk 3]
```diff
@@ -141,11 +123,16 @@ struct NCCLContextMap {
...
+  size_t size() const { return contexts_.size(); }
+
   void WaitAll() {
     for (auto &p : contexts_) {
       p.second.ctx_->Wait();
     }
   }
+
+ private:
+  static NCCLContextMap *ctx_map_ptr_;
```
Comment (typhoonzero) [inline paddle/fluid/platform/nccl_helper.h]: Is there a way not to make this singleton?

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -31,144 +31,36 @@
...
+    'sequence_enumerate', 'expand', 'sequence_concat', 'scale',
+    'elementwise_add', 'elementwise_div', 'elementwise_sub', 'elementwise_mul',
+    'elementwise_max', 'elementwise_min', 'elementwise_pow',
+    'uniform_random_batch_size_like', 'gaussian_random', 'sampling_id',
+    'gaussian_random_batch_size_like', 'sum', 'slice', 'shape', 'logical_and',
+    'logical_or', 'logical_xor', 'logical_not', 'clip', 'clip_by_norm', 'mean',
+    'mul', 'sigmoid_cross_entropy_with_logits', 'maxout', 'space_to_depth',
+    'affine_grid', 'sequence_reverse', 'affine_channel', 'similarity_focus',
+    'hash', 'grid_sampler', 'log_loss', 'add_position_encoding',
+    'bilinear_tensor_product', 'sync_fc'
```
Comment (typhoonzero) [inline python/paddle/fluid/layers/nn.py]: seems the style check have some conflicts?

File: paddle/fluid/platform/port.h [hunk 5]
```diff
@@ -14,10 +14,10 @@
 
 #pragma once
 
+#include <time.h>
 #include <cstdio>
 #include <stdexcept>
 
-#include <time.h>
```
Comment (gavin1332) [inline paddle/fluid/platform/port.h:20]: Is it necessary to reorder the time.h?

Comment (Yancey0623) [reply paddle/fluid/platform/port.h:20]: Just used to pass the pre-commit check.

**Pull request closed without merging**